Lib/functools.py

```diff
@@ -273,43 +273,130 @@ def reduce(function, sequence, initial=_initial_missing):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+def __placeholder_init_subclass__(cls, *args, **kwargs):
+    raise TypeError(f"type '{cls.__name__}' is not an acceptable base type")
+
+Placeholder = type(
+    'PlaceholderType',
+    (__PlaceholderTypeBase,),
+    {'__slots__': (), '__init_subclass__': __placeholder_init_subclass__}
+)()
+
+
+def _partial_prepare_new(cls, func, args, keywords):
+    if args:
+        if args[-1] is Placeholder:
+            raise TypeError("trailing Placeholders are not allowed")
+        phcount = args.count(Placeholder)
+    else:
+        phcount = 0
+    if isinstance(func, cls):
+        pto_args = func.args
+        pto_phcount = func.placeholder_count
+        # merge args with args of `func` which is `partial`
+        if pto_phcount and args:
+            tot_args = list(pto_args)
+            nargs = len(args)
+            pos = j = 0
+            end = nargs if nargs < pto_phcount else pto_phcount
+            while j < end:
+                pos = tot_args.index(Placeholder, pos)
+                tot_args[pos] = args[j]
+                pos += 1
+                j += 1
+            if pto_phcount < nargs:
+                tot_args.extend(args[pto_phcount:])
+            phcount += pto_phcount - end
+            args = tuple(tot_args)
+        else:
+            phcount += pto_phcount
+            args = func.args + args
+        keywords = {**func.keywords, **keywords}
+        func = func.func
+    return func, args, keywords, phcount
+
+
+def _partial_prepare_call(self, args, keywords):
+    pto_phcount = self.placeholder_count
+    pto_args = self.args
+    if pto_phcount:
+        n = len(args)
+        if n < pto_phcount:
+            raise TypeError(
+                "missing positional arguments "
+                "in 'partial' call; expected "
+                f"at least {pto_phcount}, got {n}")
+        pto_args = list(pto_args)
+        pos = j = 0
+        while j < pto_phcount:
+            pos = pto_args.index(Placeholder, pos)
+            pto_args[pos] = args[j]
+            pos += 1
+            j += 1
+        args = args[pto_phcount:] if n > pto_phcount else ()
+    keywords = {**self.keywords, **keywords}
+    return pto_args, args, keywords
+
+
+def _partial_repr(self):
+    cls = type(self)
+    module = cls.__module__
+    qualname = cls.__qualname__
+    args = [repr(self.func)]
+    args.extend(map(repr, self.args))
+    args.extend(f"{k}={v!r}" for k, v in self.keywords.items())
+    return f"{module}.{qualname}({', '.join(args)})"
+
+
 # Purely functional, no descriptor behaviour
 class partial:
     """New function with partial application of the given arguments
     and keywords.
     """
 
-    __slots__ = "func", "args", "keywords", "__dict__", "__weakref__"
+    __slots__ = ("func", "args", "keywords", "placeholder_count",
```

Comment:
`placeholder_count` is just an implementation detail. It is not necessary, it is only needed for some kind of optimization. And I think that it is better to pre-calculate a tuple of placeholder indices instead of just count them.

And as an implementation detail, it should not be a part of the public API. The attribute name in the Python code should be underscored, and the C code does not need to expose it at all.